### PR TITLE
[infra] Integrate with external PR preview system

### PR DESCRIPTION
**Pull request by jugglinmike**
Apologies for all the text. This is fairly involved, and I want to make sure everyone's on the same page. My main questions are

- Should we use the same GitHub-provided API access token used by the other Actions? If so, should we collapse this into a single long-running Action?
- Should we manage git refs using the git repository instead of the GitHub API?
- Should we ignore pull requests authored by downstream "export bots"?

I'm putting off writing automated tests until we iron out some of these details.

The pull request mirroring system is *not yet deployed* to https://wptpr.live. This shouldn't be merged before it is.

Commit message and design considerations follow.

---

Introduce a GitHub Action to monitor Pull Requests, storing relevant
information in the project's git repository (thus allowing the external
wptpr.live system to publish previews) and creating GitHub Deployments
(thus alerting contributors to the status of the preview). This Action
is triggered on a regular interval.

Introduce a second GitHub Action to monitor the state of the preview
system and communicate the relevant status to contributors via the Pull
Request UI. This Action is triggered for every GitHub Deployment created
in the previously-described Action.

For example, if three Pull Requests are updated, the first GitHub Action
will inspect them all. It will create GitHub Deployments only for the
"trusted" Pull Requests. A new GitHub Action will run for each of the
Deployments, polling the preview website until either the preview is
available or a timeout is reached. This Action will update the
deployment accordingly so that the author of each Pull Request author is
aware of the status of the preview site.

The following flow chart visually describes the same sequence:

                           sync
    gh-101 (trusted) --->   |
    gh-102 (untrusted) ->   |
    gh-103 (trusted) --->   |
                          .----.
                          |sync|--------+---------------------.
                          '----'        |                     |
                                 .-------------.       .-------------.
                                 |deploy gh-101|       |deploy gh-103|
                                 '-------------'       '-------------'
                                        |                     |
                                 poll for preview      poll for preview
                                        |                     |
    gh-101 <------ success ----- preview available     poll for preview
                                                              |
    gh-103 <------- error --------------------------------- timeout

---

Here's a demonstration of the basic behavior (intentionally using the account
of a non-collaborator) in the wpt-actions-test repository:

https://github.com/web-platform-tests/wpt-actions-test/pull/20

[The `GITHUB_TOKEN` Secret is the most direct way to grant "write" privileges to Actions.](https://help.github.com/en/github/automating-your-workflow-with-github-actions/virtual-environments-for-github-actions) Unfortunately, this script currently needs to trigger Actions from another Action, and [that's not possible with the `GITHUB_TOKEN`](https://help.github.com/en/github/automating-your-workflow-with-github-actions/events-that-trigger-workflows). There are two ways around this:

1. Use a user-generated token; **benefits**: easier to communicate state, easier to test, reduced risk of interference with other services due to API quota enforcement
2. Refactor to do all the work (creating one or more Deployments and polling for previews) from the same Action; **benefits**: does not require maintenance of another secret

For now, I've opted for the first approach (that's why the @wpt-pr-bot requested the Deployment in the demonstration above). The API request quota safeguard is less important if we're using an isolated access token, so we might want to simplify and remove that code if we stick with this approach.

The script creates and updates git refs using the GitHub API. If we wanted to reduce usage of the API (to avoid reaching the quota), we could perform this via git, but we'd have to then pull each preview-enabled branch.

Finally, I'm wondering if we should provide previews for Pull Requests opened by bots.

### CODE REVIEW

File: tools/ci/pr_preview.py
Comment (foolip) [inline tools/ci/pr_preview.py]: I'd like to not use a label here. When it was used, it cluttered up any PR listing quite badly, and when we're using deployments instead it ought not be needed.

File: .github/workflows/detect_pull_request_preview.yml
Comment (foolip) [inline .github/workflows/detect_pull_request_preview.yml]: Can you name this more in line with documentation.yml and manifest.yml? These strings show up in GitHub UI with limited width, so they need to be pretty short. If it's possible to share the same name for two workflows, calling both "pr-preview" and just having the jobs named differently would be ideal. That might be an error though.

File: .github/workflows/detect_pull_request_preview.yml
Comment (foolip) [inline .github/workflows/detect_pull_request_preview.yml]:
```suggestion
      # By default, the "checkout" action will attempt to check out the
```

File: .github/workflows/pull_request_previews.yml
Comment (foolip) [inline .github/workflows/pull_request_previews.yml]:
```suggestion
        # the "repo" scope, and is should be stored as a secret named
```

**Comment by foolip** [comment]
> Should we use the same GitHub-provided API access token used by the other Actions? If so, should we collapse this into a single long-running Action?

Using `DEPLOY_TOKEN` as you've done seems OK. That's already used in documentation.yml, is it already a @wpt-pr-bot token?

Looks like the current setup on polling vs. long-running is that the polling action triggers the long-running ones. Are you thinking of collapsing into the polling actions sometimes waiting around? If this is possible without creating a race condition between two actions trying to update the same deployment status, that sounds good.

> Should we manage git refs using the git repository instead of the GitHub API?

`create_ref` is using the API now. Would there be benefits to using the git CLI instead?

> Should we ignore pull requests authored by downstream "export bots"?

Yeah, let's do that for now.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: The GitHub Issue Label isn't a vestige from the prior implementation; it's the user-facing mechanism for enabling previews from untrusted forks. That's important even in a solution that uses GitHub Deployments. To me, it's an improvement over the comment-based approach offered by the w3c-test.org implementation.

We could use the label *only* for that purpose, so that only humans apply it. One drawback is project collaborators won't be able to opt out of the automatic preview (under the solution proposed here, they would do this by removing the automatically-assigned Label). Opting out is a feature of the w3c-test.org implementation, but I doubt it's a highly desired one, so maybe that's acceptable.

If we do limit the usage in this way, I'd like to change the name to "safelisted-for-preview".

Currently, the wpt.fyi system publishes previews for all refs that exist in two namespaces: `refs/prs-open/*` and `refs/prs-labeled-for-preview/*`. With this change, we'll certainly want to rework that. The simplest would be to rename the latter to `refs/prs-trusted-for-preview/*`, though that promotes the concept of "trust" in a way that couples the two projects. It would be more hygienic to work from *three* namespaces: `refs/prs-open/*`, `refs/prs-safelisted/*`, and `refs/prs-from-collaborators/*`.

So while removing automated application of the Label is fairly straightforward, it has ramifications that require a few hours of work. I'll hold off on that until there's agreement on the topic.

File: .github/workflows/detect_pull_request_preview.yml
Comment (jugglinmike) [reply .github/workflows/detect_pull_request_preview.yml]: Sure

File: .github/workflows/pull_request_previews.yml
Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: The documentation in this patch consistently references proprietary GitHub concepts with capital letters. I prefer this because it leaves us room to talk about things like "preview deployments" as distinct from the resources provided by GitHub. If you and the other maintainers aren't swayed by that, then I will remove all relevant capitalization.

Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: The documentation in this patch consistently references proprietary GitHub concepts with capital letters. I prefer this because it leaves us room to talk about things like "preview deployments" as distinct from the resources provided by GitHub. If you and the other maintainers aren't swayed by that, then I will remove all relevant capitalization.

Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: The documentation in this patch consistently references proprietary GitHub concepts with capital letters. I prefer this because it leaves us room to talk about things like "preview deployments" as distinct from the resources provided by GitHub. If you and the other maintainers aren't swayed by that, then I will remove all relevant capitalization.

**Comment by jugglinmike** [comment]
> Using `DEPLOY_TOKEN` as you've done seems OK. That's already used in documentation.yml, is it already a @wpt-pr-bot token?

Yup, but don't take my word for it. The proof is in [the history of the `gh-pages` branch](https://github.com/web-platform-tests/wpt/commits/gh-pages), where the @wpt-pr-bot is the author of every commit (well, almost every commit).

> Looks like the current setup on polling vs. long-running is that the polling action triggers the long-running ones. Are you thinking of collapsing into the polling actions sometimes waiting around? If this is possible without creating a race condition between two actions trying to update the same deployment status, that sounds good.

Deployments are associated with commits. Since the deployments will be created immediately in either approach, neither is more susceptible to race conditions.

> `create_ref` is using the API now. Would there be benefits to using the git CLI instead?

This is also true of `update_ref`. If we instead shelled out to the git CLI, then we would reduce the impact that this script has on the API request quota.

> > Should we ignore pull requests authored by downstream "export bots"?
>
> Yeah, let's do that for now.

Done

File: .github/workflows/pull_request_previews.yml
Comment (foolip) [inline .github/workflows/pull_request_previews.yml]: Is `0/5` equivalent to `*/5`? If so, I'd prefer the latter, since it's a form I've seen (and used) more.

File: .github/workflows/pull_request_previews.yml
Comment (foolip) [reply .github/workflows/pull_request_previews.yml]: I've you've done this consistently then OK, I won't hunt down all the capital letter. I think it looks like a typo and suspect I'm not alone, but let's wait for a second person to try correcting it before doing anything.

File: tools/ci/pr_preview.py
Comment (foolip) [reply tools/ci/pr_preview.py]: A "safelisted-for-preview" label that's only used to trigger mirroring where it otherwise wouldn't happen sounds good to me.

Since the refs are just to communicate to wptpr.live what it should do and it doesn't need to know why a PR should be deployed, won't two sets of refs still suffice? Just rename `refs/prs-labeled-for-preview/*` to `refs/prs-deployable/*` or something?

File: .github/workflows/pull_request_previews.yml
Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: Apologies, I didn't mean to suggest that you should find and replace the capitalization. If my rationale doesn't sway folks, then I'll certainly update the patch.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: > Since the refs are just to communicate to wptpr.live what it should do and it doesn't need to know why a PR should be deployed, won't two sets of refs still suffice?

That's true, and going further, we only technically need one ref. Something like `refs/prs-open-and-trusted/*` would be sufficient for wptpr.live to do its job.

Great minds think alike! That's how I originally designed [the prior iteration of this task](https://github.com/web-platform-tests/wpt/pull/17680). @jgraham [pushed back on that, writing](https://github.com/web-platform-tests/wpt/pull/17680#issuecomment-510000469)

> it highly couples things to the exact requirements of live preview; we only get tags on items that are both open PRs and also have the right label set. We could decouple those things and create refs for open PRs and refs for things labelled for preview.

I agreed, which is why this iteration started with two ref namespaces. My suggestion to introduce a third namespace is an application of the same rationale.

Then again, while a distinct `prs-open` ref namespace might be useful elsewhere, I'm less confident that a `prs-from-collaborators` would be (and even less so about `prs-safelisted`). I don't want to increase complexity to reduce coupling if there's no practical benefit. @jgraham do you have an opinion?

File: .github/workflows/pull_request_previews.yml
Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: Yes, it is equivalent. Updated

**Comment by zcorpan** [comment]
I removed the `.github` label per https://github.com/web-platform-tests/wpt-pr-bot/pull/105

Let's see if a new label appears...

File: tools/ci/pr_preview.py
Comment (jgraham) [reply tools/ci/pr_preview.py]: Not a strong opinion, but I think two is fine. "PRs that are open" seems like a clearly reusable thing, but trying to split the difference between "PRs that are trusted because they're from collaborators" and "PRs that are trusted because they were marked as such" seems like it might be overkill in terms of enabling other consumers.

Comment (jugglinmike) [reply tools/ci/pr_preview.py]: Thanks, @jgraham. I've updated this accordingly. To summarize: the GitHub Issue label in use is "safelisted-for-preview", and it is only referenced by this script (that is, never introduced) in order to determine when a submission from a non-Collaborator should be mirrored. The underlying ref namespace is `refs/prs-trusted-for-preview/*`.

...I think. This is arduous to verify because the patch still lacks automated tests. Now that three members of the "wpt core team" have participated in this discussion, and no one has objected, I think it's prudent to write some tests. I'll plan to get started on that tomorrow.

**Comment by jugglinmike** [comment]
Okay, tests are up. Also, support for Python 3--tox is great!

**Comment by foolip** [comment]
@jugglinmike there are still CI failures, can you take a look?

**Comment by jugglinmike** [comment]
@foolip I've corrected the Windows failures, but now there are problems with the Taskcluster integration. Closing and re-opening this pull request [as James suggested](https://github.com/orgs/web-platform-tests/teams/reviewers/discussions/3) didn't resolve the problem. Any ideas about what I should try next?

**Comment by foolip** [comment]
With Taskcluster it’s often necessary to rebase on a more recent master. @jgraham is that the case here and do your instructions need to be updated?

**Comment by jgraham** [comment]
Oh yeah, I guess this won't have the required changes to `.taskcluster.yml`.

**Comment by jugglinmike** [comment]
Okay, I've rebased and force-pushed the result. [The original versoin of this branch is available here.](https://github.com/web-platform-tests/wpt/pull/17189#issuecomment-552561665)

File: .github/workflows/pull_request_previews.yml
Comment (foolip) [inline .github/workflows/pull_request_previews.yml]: Can we do a shallow checkout to make this faster, given how often it'll run?

File: tools/ci/pr_preview.py
Comment (foolip) [inline tools/ci/pr_preview.py]:
```suggestion
    'autofoolip', 'chromium-wpt-export-bot', 'moz-wptsync-bot', 'servo-wpt-sync'
```

File: tools/ci/pr_preview.py
Comment (foolip) [inline tools/ci/pr_preview.py]:
```suggestion
LABEL = 'safe for preview'
```

Mostly because it's shorter. It seems like most labels either match directory names, are on the form "foo:bar-baz" or "human readable string". I didn't count though, and did see exceptions.

**Comment by foolip** [comment]
@stephenmcgruer @Hexcles would either of you be up for reviewing this?

File: .github/workflows/pull_request_previews.yml
Comment (jugglinmike) [reply .github/workflows/pull_request_previews.yml]: Sure

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: It's more accurate, as well, since adding the label *is* what makes a pull request "safelisted."

**Comment by stephenmcgruer** [comment]
Happy to review, but at BlinkOn until next week. Wouldn't be able to look at this till Monday, sorry.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [inline tools/ci/pr_preview.py]: Nit; the generic name `LABEL` and `HasLabel` functions mislead me a bit when reading - it required me to come up to this definition and see that the label is the one that marks it as safe to mirror. Maybe just renaming `HasLabel` to `HasMirroringLabel` would suffice to remove my confusion.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [inline tools/ci/pr_preview.py]: Does this mean that if a PR with the 'safe for preview' label is mirrored (creating a prs-trusted-for-preview/XYZ ref), that ref will never be removed even after the PR is closed? If that's intended, I think maybe I don't understand the flow here.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [inline tools/ci/pr_preview.py]: Does the 'gh-' prefix here have meaning to the GitHub APIs, or is it to uniquely identify PR preview deployments? If the latter, we may want to use something more specific.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [inline tools/ci/pr_preview.py]: Nit; we will never log non-successful status codes. (I assume, given `raise_for_status` above)

**Review by stephenmcgruer — APPROVED** [review]
Generally LGTM. A few comments/questions, but nothing blocking I think.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: I considered parameterizing the label name, but that seemed a little silly given that we only expect one value. Renamed the function as you requested.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: That's correct; trust is independent from the open/closed state of the pull request. [The mirroring system only publishes previews for pull requests that have both refs](https://github.com/bocoup/web-platform-tests.live/blob/95770444285e86ec5851f30a8a5395ed1d79216b/src/mirror-pull-requests.sh#L41), so removing either will cancel the preview.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: Good idea.

Making this change surfaced an integration bug. Since I wrote this script, [we changed the URLs generated by the preview server](https://github.com/bocoup/web-platform-tests.live/pull/23), but I forgot to accommodate that change in the logic that detects preview status. I've corrected that, too.

File: tools/ci/pr_preview.py
Comment (jugglinmike) [reply tools/ci/pr_preview.py]: The error message will describe the status, but a caller could handle or override the exception. I'll swap these two lines, just to be safe.

**Comment by jugglinmike** [comment]
Thanks for the review, @stephenmcgruer! Unfortunately, unrelated infrastructure failures keep getting in the way of the validation for this patch. I've pushed up another empty commit; hopefully, the third time's the charm (though I won't be around to see until tomorrow).

**Comment by stephenmcgruer** [comment]
Dammit, I totally missed "The pull request mirroring system is not yet deployed to https://wptpr.live. This shouldn't be merged before it is.".

@jugglinmike Unlikely you're still awake, but is the PR mirroring system deployed to wptpr.live yet?

**Comment by Hexcles** [comment]
It has been deployed, but looks like Mike has not verified it end to end?

**Comment by stephenmcgruer** [comment]
Well... this is one way to verify it end to end. I'm intrigued, how about we let this new action run at least once and see what happens?

**Comment by jugglinmike** [comment]
Robert's right on both counts.

Something has certainly happened. http://wptpr.live currently offers a slew of unexpected html documents at the top level. Investigating presently

**Comment by stephenmcgruer** [comment]
Also:

```
.github/workflows/pull_request_previews.yml
Invalid Workflow File

DETAILS
yaml: line 4: did not find expected alphabetic or numeric character
```

**Comment by stephenmcgruer** [comment]
Btw, Robert and I looked at http://wptpr.live/ the other day (before this) and it looked much like it looks currently (in terms of those unexpected documents)

**Comment by jugglinmike** [comment]
> Btw, Robert and I looked at http://wptpr.live/ the other day (before this) and it looked much like it looks currently (in terms of those unexpected documents)

Yeah, given the nature of the problem being reported by GitHub, those files couldn't have been created in response to the merging of this pull request.

Here's a fix for the syntax error: https://github.com/web-platform-tests/wpt/pull/20347

**Comment by stephenmcgruer** [comment]
With the syntax error fixed, we have had a pr-preview action trigger successfully. It hit https://api.github.com/search/issues?q=repo:web-platform-tests/wpt+is:pr+updated:>2019-11-20T20:03:09Z and found 0 pull requests.

**Comment by jugglinmike** [comment]
The subsequent run demonstrates slightly more of the intended behavior:

    ./tools/ci/pr_preview.py --host https://api.github.com --github-project web-platform-tests/wpt synchronize --window 480
    INFO:__main__:Issuing request: GET https://api.github.com/rate_limit
    INFO:__main__:Response status code: 200
    INFO:__main__:Limit for "search" resource: 30/30
    INFO:__main__:Searching for Pull Requests updated since 2019-11-20T20:09:07Z
    INFO:__main__:Issuing request: GET https://api.github.com/search/issues?q=repo:web-platform-tests/wpt+is:pr+updated:>2019-11-20T20:09:07Z
    INFO:__main__:Response status code: 200
    INFO:__main__:Found 2 Pull Requests
    INFO:__main__:Processing Pull Request #19838
    INFO:__main__:Pull Request should not be mirrored
    INFO:__main__:Processing Pull Request #19632
    INFO:__main__:Pull Request should not be mirrored

Pull request number 19838 is closed, and pull request number 19632 was opened by a non-collaborator

**Comment by stephenmcgruer** [comment]
That issue is a bit hairy, so I'm waiting for a 'real' PR to drop a `safe for preview` label on (or lmk if you don't want that done yet)

**Comment by stephenmcgruer** [comment]
Ah, I see you deliberately triggered it on PR 20239. Fingers crossed!

**Comment by stephenmcgruer** [comment]
http://wptpr.live/20239/ exists! \o/

**Comment by jugglinmike** [comment]
[The script successfully created a deployment](https://github.com/web-platform-tests/wpt/commit/70f36aea7d836d051a38c34e1a3d3dd15b4bca45/checks?check_suite_id=321002227), and [the corresponding pull request was updated to describe an "in progress" deployment](https://github.com/web-platform-tests/wpt/pull/20239). wptpr.live recognized the new tags and mirrored as anticipated:

    $ curl http://wptpr.live/.git/worktrees/20239/HEAD
    c2cf7ba8a6730e847724d6d8c718ab4a2eafc0ce
    $ curl --silent http://wptpr.live/20239/LICENSE.md | head -n1
    # The 3-Clause BSD License

...but the corresponding `pr-preview-detect` workflow did not run. Becuase of that, the GitHub UI for the pull request still lists the deployment as "pending" and does not include a link to the preview.

I will continue to investigate and report back here.

**Comment by sideshowbarker** [comment]
> Something has certainly happened. http://wptpr.live currently offers a slew of unexpected html documents at the top level. Investigating presently

That happens sometimes at http://w3c-test.org/ too. In fact that’s the same state it’s in right now. When it happens, I’ve just gone into the environment on the host and manually deleted the files. I’ve never been able to discern the cause (though I’ve not really ever seriously investigated it).

**Pull request merged**
